homeassistant/components/wyoming/conversation.py

```diff
@@ -0,0 +1,207 @@
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.util import ulid
+
+from .const import DOMAIN
+from .data import WyomingService
+from .error import WyomingError
+from .models import DomainDataItem
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config_entry: ConfigEntry,
+    async_add_entities: AddEntitiesCallback,
+) -> None:
+    """Set up Wyoming conversation."""
+    item: DomainDataItem = hass.data[DOMAIN][config_entry.entry_id]
+    async_add_entities(
+        [
+            WyomingConversationEntity(config_entry, item.service),
+        ]
+    )
+
+
+class WyomingConversationEntity(
+    conversation.ConversationEntity, conversation.AbstractConversationAgent
+):
+    """Wyoming conversation agent."""
+
+    _attr_has_entity_name = True
+
+    def __init__(
+        self,
+        config_entry: ConfigEntry,
+        service: WyomingService,
+    ) -> None:
+        """Set up provider."""
+        super().__init__()
+
+        self.entry = config_entry
+        self.service = service
+
+        self._intent_service: IntentProgram | None = None
+        self._handle_service: HandleProgram | None = None
+
+        for maybe_intent in self.service.info.intent:
+            if maybe_intent.installed:
+                self._intent_service = maybe_intent
+                break
+
+        for maybe_handle in self.service.info.handle:
+            if maybe_handle.installed:
+                self._handle_service = maybe_handle
+                break
+
+        model_languages: set[str] = set()
+
+        if self._intent_service is not None:
+            for intent_model in self._intent_service.models:
+                if intent_model.installed:
+                    model_languages.update(intent_model.languages)
+
+            self._attr_name = self._intent_service.name
+            self._attr_supported_features = (
+                conversation.ConversationEntityFeature.CONTROL
+            )
+        elif self._handle_service is not None:
+            for handle_model in self._handle_service.models:
+                if handle_model.installed:
+                    model_languages.update(handle_model.languages)
+
+            self._attr_name = self._handle_service.name
+        else:
+            raise ValueError("No intent or handle service")
+
+        self._supported_languages = list(model_languages)
+        self._attr_unique_id = f"{config_entry.entry_id}-conversation"
+
+    async def async_added_to_hass(self) -> None:
+        """When entity is added to Home Assistant."""
+        await super().async_added_to_hass()
+        conversation.async_set_agent(self.hass, self.entry, self)
+
+    async def async_will_remove_from_hass(self) -> None:
+        """When entity will be removed from Home Assistant."""
+        conversation.async_unset_agent(self.hass, self.entry)
+        await super().async_will_remove_from_hass()
```

Comment:
Only do this if the conversation agent was previously available for backwards compatibility. I don't think this is necessary because this is the first time Wyoming exposes a conversation agent?

Reply:
This appears to be quite necessary, actually :smile: 
I had just removed the migration part.

Reply:
Can you explain why you think it is necessary? Creating a conversation entity is the modern way. Setting agents was the old way that we shuoldn't add to new platforms.

Reply:
When I removed it, the agent no longer worked. Should we be calling `async_set_agent` automatically in the entity base class?

Reply:
No, that should never be called anymore. It is no longer used. Users should only target the entity ID.